### PR TITLE
Feature/retestid

### PR DESCRIPTION
**Pull request by roesslerj**

### CODE REVIEW

File: src/main/java/de/retest/ui/descriptors/Element.java [hunk 1]
```diff
@@ -25,6 +26,9 @@
 
 	protected static final long serialVersionUID = 2L;
 
+	@XmlAttribute
+	protected final String retestId;
```
Comment (beatngu13) [inline src/main/java/de/retest/ui/descriptors/Element.java]: Isn't this a (massive) breaking change as we change all ctors?

File: src/test/java/de/retest/ui/descriptors/DefaultRetestIdProviderTest.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+		final DefaultRetestIdProvider cut = new DefaultRetestIdProvider();
+
+		// for what you would actually call a text
+		final IdentifyingAttributes ident = createIdentAttributes(
+				"This is some very long sentence, that could be in a link text, or in some paragraph, and really is to long to be used as id." );
+		final String retestId = cut.getRetestId( ident );
+		assertThat( retestId.length() ).isLessThan( 20 );
+
+		// and cut should still be unique!
+		assertThat( retestId.length() ).isNotEqualTo( cut.getRetestId( ident ) );
```
Comment (beatngu13) [inline src/test/java/de/retest/ui/descriptors/DefaultRetestIdProviderTest.java]: Doesn't this compare integer (`retestId.length()`) with string (`cut.getRetestId( ident )`)?

File: src/test/java/de/retest/ui/descriptors/DefaultRetestIdProviderTest.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	public void too_long_words_should_be_cut() {
+		final DefaultRetestIdProvider cut = new DefaultRetestIdProvider();
+
+		// but also for single words
+		final IdentifyingAttributes ident = createIdentAttributes( "supercalifragilisticexpialidocious" );
+		final String retestId = cut.getRetestId( ident );
+		assertThat( retestId.length() ).isLessThan( 20 );
+
+		// and cut should still be unique!
+		assertThat( retestId.length() ).isNotEqualTo( cut.getRetestId( ident ) );
```
Comment (beatngu13) [inline src/test/java/de/retest/ui/descriptors/DefaultRetestIdProviderTest.java]: Same here: integer vs. string.

File: src/test/java/de/retest/ui/descriptors/DefaultRetestIdProviderTest.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	}
+
+	public IdentifyingAttributes createIdentAttributes( final String text ) {
```
Comment (beatngu13) [inline src/test/java/de/retest/ui/descriptors/DefaultRetestIdProviderTest.java]: Should be `private`.

File: src/test/java/de/retest/ui/descriptors/DefaultRetestIdProviderTest.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+		final Collection<Attribute> attributes = createList( fromString( "/HTML/DIV" ), "DIV" );
+		attributes.add( new StringAttribute( "text", text ) );
+		final IdentifyingAttributes ident = new IdentifyingAttributes( attributes );
```
Comment (beatngu13) [inline src/test/java/de/retest/ui/descriptors/DefaultRetestIdProviderTest.java]: Can be inlined.

File: src/test/java/de/retest/ui/descriptors/DefaultRetestIdProviderTest.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+	@Test
+	public void should_always_be_unique() {
+		final DefaultRetestIdProvider cut = new DefaultRetestIdProvider();
+		final IdentifyingAttributes ident = createIdentAttributes( "a" );
+		final String retestId = cut.getRetestId( ident );
+		assertThat( retestId ).isNotEqualTo( cut.getRetestId( ident ) );
+		assertThat( retestId ).isNotEqualTo( cut.getRetestId( ident ) );
+		assertThat( retestId ).isNotEqualTo( cut.getRetestId( ident ) );
+		assertThat( retestId ).isNotEqualTo( cut.getRetestId( ident ) );
```
Comment (beatngu13) [inline src/test/java/de/retest/ui/descriptors/DefaultRetestIdProviderTest.java]: Same check 4 times?

File: src/test/java/de/retest/ui/descriptors/DefaultRetestIdProviderTest.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+	@Test( expected = NullPointerException.class )
+	public void null_should_give_exception() {
+		final DefaultRetestIdProvider cut = new DefaultRetestIdProvider();
+		cut.getRetestId( null );
+	}
+
+	@Test( expected = NullPointerException.class )
+	public void null_path_should_give_exception() {
+		final DefaultRetestIdProvider cut = new DefaultRetestIdProvider();
+		cut.getRetestId( create( null, "DIV" ) );
```
Comment (beatngu13) [inline src/test/java/de/retest/ui/descriptors/DefaultRetestIdProviderTest.java]: This is not behavior of the CUT, the `NullPointerException` is thrown by `PathAttribute`. The CUT has no check for a `null` path.

File: src/test/java/de/retest/ui/descriptors/ElementTest.java [hunk 8]
```diff
@@ -18,7 +18,8 @@
 	public void toString_returns_UniqueCompIdentAttributes_toString() throws Exception {
 		final IdentifyingAttributes compIdentAttributes =
 				IdentifyingAttributes.create( fromString( "Window/path/Component" ), java.awt.Component.class );
-		assertThat( new Element( compIdentAttributes, null ).toString() ).isEqualTo( compIdentAttributes.toString() );
+		assertThat( new Element( "", compIdentAttributes, null ).toString() )
```
Comment (beatngu13) [inline src/test/java/de/retest/ui/descriptors/ElementTest.java]: Isn't this a flaw that we allow IDs to be empty? Should be like a variable, at least one alphanumeric char (IMHO).

File: src/test/java/de/retest/ui/descriptors/DefaultRetestIdProviderTest.java (hunk 6 above)
Comment (roesslerj) [reply src/test/java/de/retest/ui/descriptors/DefaultRetestIdProviderTest.java]: getRetestId should return a new unique ID every time...

File: src/test/java/de/retest/ui/descriptors/DefaultRetestIdProviderTest.java (hunk 7 above)
Comment (roesslerj) [reply src/test/java/de/retest/ui/descriptors/DefaultRetestIdProviderTest.java]: True. But this behavior should be checked here non the less.

File: src/test/java/de/retest/ui/descriptors/ElementTest.java (hunk 8 above)
Comment (roesslerj) [reply src/test/java/de/retest/ui/descriptors/ElementTest.java]: Good catch.

**Comment by roesslerj** [comment]
I adressed your requested changes.

**Pull request merged**